File: Ontology/DataType.cs
using System;

namespace Ontology
{
    public class Datatype:
        Entity,
        IDatatype
    {
        public Datatype(
            IOntology ontology,
            string    localName
            ) : base(
                ontology,
                localName)
        {
        }

        bool IDataRange.HasMember(
            object value
            )
        {
            throw new NotImplementedException();
        }
    }

    public class Datatype<T>:
        BuiltIn,
        IDatatype
    {
        public Datatype(
            string prefixName,
            string localName
            ) : base(
                ReservedVocabulary.StandardPrefixNames[prefixName],
                localName)
        {
        }

        bool IDataRange.HasMember(
            object value
            )
        {
            return value is T;
        }
    }
}

File: Web/ClientApp/src/app/Ontology/Datatype.ts
import { Entity } from "./Entity";
import { IDatatype, IDatatypeSelector } from "./IDatatype";
import { IOntology } from "./IOntology";

export class Datatype
    extends Entity
    implements IDatatype
{
    constructor(
        ontology : IOntology,
        localName: string
        )
    {
        super(
            ontology,
            localName);
    }

    HasMember(
        value: object
        ): boolean
    {
        throw new Error("Method not implemented.");
    }

    Select<TResult>(
        selector: IDatatypeSelector<TResult>
        ): TResult
    {
        return selector.Datatype(this);
    }
}
